0027-safely-recommend-actions.adoc

```diff
@@ -0,0 +1,173 @@
+
+----
+1. IF refund is available RETURN refund
+2. IF there is time to transition to next AND there is time for the subsequent action RETURN next
+3. IF have funded RETURN wait_for_refund
+4. RETURN abort
+----
+
+Note that step 2 is recursive, the base case being the final stage of the annotated state diagram - done (see below).
+
+=== Explanation
+
+Various events must occur in order to transition from one state to the next. We can identify these events and annotate the state diagram accordingly. If we are able to calculate the time of each of these events we are able to fully determine the period of each transition.
+
+We define the Time To Final to be a part of the transition to a state e.g., the alpha funded state is not reached until the alpha fund transaction has become final.
+
+There exists a transition diagram for each actor. Each actor progresses along their own state diagram but must be able to predict the state progression of the counterparty in order to determine whether progression is possible for the counterparty.
+
+.Annotated state diagram for Alice:
+----
+(Create) TTA_A TTI_AF TTF_A (alpha funded) TTA_B TTI_BF TTF_B (beta funded) TTA_A TTI_BR TTC_B (beta redeemed) (done)
+----
+
+.Annotated state diagram for Bob:
+----
+(Create) TTA_A TTI_AF TTF_A (alpha funded) TTA_B TTI_BF TTF_B (beta funded) TTA_A TTI_BR (beta redeemed) TTA_B TTI_AR TTC_A (alpha redeemed) (done)
+----
+
+For a swap to continue, at all times, each actor must be able to transition to 'done' before the expiry time governing their redeem action elapses.
+
+.Key:
+* (state): states are shown inside parenthesis
+* `TTA_A`: `TTA` for Alice
+* `TTA_B`: `TTA` for Bob
+* `TTI_AF`: `TTI` for Alpha Fund transaction
+* `TTI_BF`: `TTI` for Beta Fund transaction
+* `TTI_AR`: `TTI` for Alpha Redeem transaction
+* `TTI_BR`: `TTI` for Beta Redeem transaction
+* `TTF_A`: `TTF` for Alpha ledger
+* `TTF_B`: `TTF` for Beta ledger
+
+.Note
+* `TTI_AR` is not actually essential for Bob, we assume Bob waits for Alice's redeem transaction to be included in a block, technically by monitoring unconfirmed transactions (e.g. watching the Bitcoin mempool) `TTI_AR` could be removed. This in no way effects the security of the protocol.
+
+
+== Calculating Events
+
+=== Time to Act
+
+Time To Act should allow for network congestion and expected response of the user.
+
+* `TTA_B` is defined to be 15 minutes
+* `TTA_A` is defined to be 60 minutes
+
+=== Time To Include
+
+(TODO: This section is incomplete)
+
+This is a common task within the ecosystem of each blockchain, as such we will do as is typical for each. We dynamically determine the suggested fee after defining `P` and N.
+
+.Values (Bitcoin/Ethereum)
+* P: probability ???
+* N: number of blocks ???
+
+=== Time To Final
+
+We consider a transaction final once `X` number of blocks have been accepted by the network since the transaction was included.
+
+.We define `X` to be
+* Bitcoin: 6
+* Ethereum: 12
```

Comment:
This is not enough if it should be similar to Bitcoin :)

6 BTC blocks are effectively one hour, Ethereum mines 210 blocks in an hour :D

Reply:
Is there a reason it should it be similar to Bitcoin? I just did a web search on what number of confirmations folks in the Ethereum ecosystem use and came up with 12.